Resend transactions fix and refactor

fedejinich opened the pull request:
# Resend Transactions

## Goals
- reBroadcast transactions
- Remove TransactionNodeInfo which was storing a lot of transactions (increasing memory use) 

## Context

There are some cases where you want to resend transactions to activePeers for example when, they have expired, or mining nodes have been restarted, or if there were a bug where tx were lost like the [previous bug](https://github.com/rsksmart/rskj/issues/1127).

When @nicops  [fixed](https://github.com/rsksmart/rskj/pull/1253) [this issue](https://github.com/rsksmart/rskj/issues/1127), he discover that it wasn't posible to rebroadcast transactions to activePeers because `TransactionNodeInformation` was blocking it. Then we decided to remove that class because that blocking was no longer needed.

This is the actual flow (before this PR) when a transaction arrives to the node:
- RPC flows:
	- `PersonalModuleWalletEnabled.sendTransaction` -> `EthereumImpl.submitTransaction` -> `TransactionPoolImpl.addTransaction`
	 - `Web3EthModule.eth_sendRawTransaction` -> `EthModuleTransactionBase.sendRawTransaction` -> `TransactionPoolImpl.addTransaction`
	 - `Web3EthModule.eth_sendTransaction` -> `EthModuleTransactionBase.sendTransaction` -> `TransactionPoolImpl.addTransaction`
- Peer node flow:
    - `MessageVisitor.apply(TransactionsMessage)` -> `TransactionGateway.receiveTransactionsFrom` -> `TransactionPool.addTransactions` -> `TransactionPool.addTransaction`

## Implemented

- Removed `TransactionNodeInfo` class (and tests)
- Refactored `TransactionGateway` class
    - `receiveTransactionsFrom` now broadcasts transactions after adding them to `TransactionPool`
    - Removed a nonsense listener: `OnPendingTransactionsReceivedListener`
- Incoming transactions should only be added to TxPool via TxGateway: replaced where we used to call `transactionPool.addTransaction` with `transactionGateway.receiveTransaction/sFrom` (because invoking `addTransaction` was also producing a `onPendingTransactionsReceived` & `onTransactionPoolChanged` event )
- Removed `@Ignore` annotation from ChanelManagerImplTests 

Note: `onPendingTransactionsReceived` was being consumed from `TransactionGateway` which then used to call `channelManager.broadcastTransaction`. Apart from this there is another consumer from this event which is not related with this PR.

fedejinich commented:
added some comments at `TransactionPool`

patogallaiovlabs commented on rskj-core/src/main/java/co/rsk/net/TransactionGateway.java:
```diff
@@ -18,66 +18,41 @@
-
-                newNodes.forEach(nodeID -> transactionNodeInformation.addTransactionToNode(txHash, nodeID));
-            }
+    public TransactionPoolAddResult receiveTransactionFrom(Transaction transaction) {
```
why the suffix "From"?

fedejinich replied:
removed "from" suffix

patogallaiovlabs commented on rskj-core/src/main/java/org/ethereum/core/TransactionPoolAddResult.java:
```diff
@@ -42,11 +48,15 @@ public void ifTransactionWasNotAdded(Consumer<String> errorConsumer) {
 
     public static TransactionPoolAddResult withError(String errorMessage) {
-        return new TransactionPoolAddResult(false, errorMessage);
+        return new TransactionPoolAddResult(false, errorMessage, null);
```
This could cause a NPE in the check "if(!result.getTransactionsAdded().isEmpty()) {" L53 TransactionGateway.

fedejinich replied:
fixed

patogallaiovlabs commented on rskj-core/src/main/java/org/ethereum/net/server/ChannelManager.java:
```diff
@@ -78,4 +78,6 @@
 
     boolean isAddressBlockAvailable(InetAddress address);
+
+    Set<NodeID> broadcastTransactions(@Nonnull List<Transaction> transactions, @Nonnull Set<NodeID> nodeID);
```
Be careful with the use of NodeID... most of the impl returns a copy of those. So, i wonder if it is actually cheaper to return the Node instead of NodeID.

fedejinich replied:
didn't get that. do you think it's worth it? this is just a reuse of broadcastTransaction which has been there before this PR

fedejinich commented:
the weird thing is that there was an `@Ignore` for ChannelManagerImplTests. I've removed that

fedejinich commented:
pipeline:run

fedejinich commented:
pipeline:run

fedejinich commented:
squashed commits

nicops commented:
> When @nicops fixed issue, he discover that there was another step to complete that fix: When a reorganisation was produced the node wasn’t resending those transactions to active peers, because we used to kept track of transactions that were already sent to other nodes.

Just to clarify, the fix I did for reorganization is more or less independent of this. Yes, when a reorganization is added, the node isn't resending those transactions. But neither is on this PR, because it doesn't go through the tx gateway. And that's fine because you don't need to resend them, it is expected other nodes already have them.

Instead, this change would apply for *any* scenario you want to resend a tx if, for example, they have expired, or mining nodes have been restarted, or if there were a bug where tx were lost like the previous bug.

fedejinich commented:
Last updates:
- Only broadcasting pending transactions instead of broadcasting queued and pending transactions
- Avoid caching Message.Transaction at NodeMessageHandler.receivedMessage -> this allows resending expired txs

fedejinich commented:
Now It's also posible to share the same logic at `addTransaction/addTransactions` at `TransactionPool`

nicops commented on rskj-core/src/test/java/co/rsk/core/bc/TransactionPoolImplTest.java:
```diff
@@ -512,8 +514,10 @@ public void addTwiceAndGetQueuedTransaction() {
-        Assert.assertFalse(result.transactionWasAdded());
-        result.ifTransactionWasNotAdded(msg -> Assert.assertEquals("queued transaction with same hash already exists", msg));
+        Assert.assertFalse(result.transactionsWereAdded());
+        if(!result.transactionsWereAdded()) {
```
You should remove this if, it is confusing, you already asserted it was false the line above

fedejinich replied:
yeah it sounds weird but in this way i'm asserting error message

fedejinich replied:
removed unnecesary if

nicops commented on rskj-core/src/test/java/co/rsk/core/bc/TransactionPoolImplTest.java:
```diff
@@ -557,8 +561,10 @@ public void checkTxWithSameNonceIsRejected() {
-        Assert.assertFalse(result.transactionWasAdded());
-        result.ifTransactionWasNotAdded(msg -> Assert.assertEquals("gas price not enough to bump transaction", msg));
+        Assert.assertFalse(result.transactionsWereAdded());
+        if(!result.transactionsWereAdded()) {
```
ditto

fedejinich replied:
removed unnecessary if

nicops commented on rskj-core/src/test/java/co/rsk/core/bc/TransactionPoolImplTest.java:
```diff
@@ -583,8 +589,10 @@ public void checkTxWithHighGasLimitIsRejected() {
-        Assert.assertFalse(result.transactionWasAdded());
-        result.ifTransactionWasNotAdded(msg -> Assert.assertEquals("transaction's gas limit of 3000001 is higher than the block's gas limit of 3000000", msg));
+        Assert.assertFalse(result.transactionsWereAdded());
+        if(!result.transactionsWereAdded()) {
```
ditto

fedejinich replied:
removed unnecessary if

nicops commented on rskj-core/src/test/java/co/rsk/core/bc/TransactionPoolImplTest.java:
```diff
@@ -802,8 +825,8 @@ public void addTwoTransactionsOutOfOrderInNonceUsingAddTransaction() {
-        Assert.assertTrue(transactionPool.addTransaction(tx2).transactionWasAdded());
-        Assert.assertTrue(transactionPool.addTransaction(tx1).transactionWasAdded());
+        Assert.assertTrue(transactionPool.addTransaction(tx2).transactionsWereAdded());
+        Assert.assertTrue(transactionPool.addTransaction(tx1).transactionsWereAdded());
```
I would make more assertions here about the return value of each addTransaction. Check that the first one returns ok queued with tx2,  the latter one should add ok pending with the list of the two tx in the right order (first tx1, then tx2).

nicops replied:
the test below (addTwoTransactionsOutOfOrderInNonceUsingAddTransactions), you didn't modify it on this PR, but should have assertions done to the return value as well

fedejinich replied:
also this `addTransaction` only retrieves an internal `TransactionPoolAddResult` instead of returning internal result+successors (successors are also in pending state now)

fedejinich replied:
I'll raise another issue to refactor those two methods into one

fedejinich replied:
#1286 https://github.com/rsksmart/rskj/issues/1286

nicops replied:
:ok_hand: :ok_hand: :ok_hand:

nicops commented on rskj-core/src/main/java/co/rsk/rpc/modules/eth/EthModuleTransactionBase.java:
```diff
@@ -64,12 +67,15 @@ public synchronized String sendTransaction(Web3.CallArguments args) {
             }
 
-            synchronized (transactionPool) {
+            synchronized (transactionGateway) {
```
I don't think this is correct. This method derives the account nonce based on the line right below this one. To do that consistently it needs to synchronized over the `transactionPool` because it is accessing it as you can see in the line right below this one.

fedejinich replied:
got it

fedejinich replied:
now it synchronises with txPool instead of txGateway

fedejinich commented:
In [last commit](https://github.com/rsksmart/rskj/commit/f6216e6df378c957096921d8450c505c0ef4819c) 
- I've added more detailed test cases for `TransactionPool` 
- Fixed a bug at `addTransaction`: it was only returning one pending transaction instead of returning that pending transaction and its successors

fedejinich commented:
Also spotted a posible refactor #1286

fedejinich commented:
In [last commit](https://github.com/rsksmart/rskj/commit/12839722581405c1d3446298a32e880c172b0481) 
- I've added more detailed test cases for `TransactionPool` 
- Synchronised `transactionPool` instead of `transactionGateway` at `EthModuleTransactionBase.sendTransaction`
- Fixed a bug at `addTransaction`: it was only returning one pending transaction instead of returning that pending transaction and its successors

fedejinich commented:
pipeline:run

fedejinich commented on rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java:
```diff
@@ -118,7 +119,7 @@ public void postMessage(Peer sender, Message message) {
         Keccak256 encodedMessage = new Keccak256(HashUtil.keccak256(message.getEncoded()));
         if (!receivedMessages.contains(encodedMessage)) {
-            if (message.getMessageType() == MessageType.BLOCK_MESSAGE || message.getMessageType() == MessageType.TRANSACTIONS) {
+            if (message.getMessageType() == MessageType.BLOCK_MESSAGE) {
```
well this was a team decision, the idea was based on the case "when a transaction expires". in the old way we couldn't even process that resented transaction (despite expiration time)

fedejinich replied:
```
But now, this logic helps a lot to discard incoming message that came many times (ie, now it's easy to see the transaction message come from many peers, even before the first transaction message could be processed)
```
what do you think about this @juli @donequis ?

donequis replied:
Does it still help? These are the first lines of the `internalAddMethod`

```
        if (pendingTransactions.hasTransaction(tx)) {
            return TransactionPoolAddResult.withError("pending transaction with same hash already exists");
        }

        if (queuedTransactions.hasTransaction(tx)) {
            return TransactionPoolAddResult.withError("queued transaction with same hash already exists");
        }
```

fedejinich commented on rskj-core/src/main/java/co/rsk/net/TransactionGateway.java:
```diff
@@ -18,66 +18,41 @@
  * have it.
  */
-public class TransactionGateway implements InternalService {
+public class TransactionGateway {
```
💯

fedejinich commented on rskj-core/src/main/java/co/rsk/core/bc/TransactionPoolImpl.java:
```diff
@@ -159,23 +159,23 @@ private RepositorySnapshot getCurrentRepository() {
 
     private List<Transaction> addSuccessors(Transaction tx) {
-        List<Transaction> added = new ArrayList<>();
+        List<Transaction> pendingTransactionsToBeAdded = new ArrayList<>();
```
yes you're right!

fedejinich commented:
In [this commit](https://github.com/rsksmart/rskj/commit/bd15017245dad912d1c58281d395f459af9bed0d) 
- I've added more detailed test cases for `TransactionPool` 
- Synchronised `transactionPool` instead of `transactionGateway` at `EthModuleTransactionBase.sendTransaction`
- Fixed a bug at `addTransaction`: it was only returning one pending transaction instead of returning that pending transaction and its successors

NOTE: I've added one test case to `NodeMessageHandlerTest` in [this](https://github.com/rsksmart/rskj/pull/1266/commits/a615c435adfce21b68a9fa05f881b9b196fbc9b8) commit

fedejinich commented:
pipeline:run

donequis commented on rskj-core/src/main/java/co/rsk/net/NodeMessageHandler.java:
```diff
@@ -127,7 +126,8 @@ private void tryAddMessage(Peer sender, Message message) {
             double score = sender.score(System.currentTimeMillis(), message.getMessageType());
 
-            if (score >= 0 && !this.queue.offer(new MessageTask(sender, message, score))) {
+            boolean notAdded = !this.queue.offer(new MessageTask(sender, message, score));
```
Messages should be offered only on non negative scores

fedejinich replied:
extracted that part into `addMessage` method
